packages/@react-aria/overlays/src/useModal.tsx

```diff
@@ -122,16 +122,25 @@ interface OverlayContainerProps extends ModalProviderProps {
  * be accessible at once.
  */
 export function OverlayContainer(props: OverlayContainerProps): React.ReactPortal {
-  let {portalContainer = document.body, ...rest} = props;
+  let {portalContainer, ...rest} = props;
+  let [currentPortalContainer, setCurrentPortalContainer] = useState(portalContainer);
 
   React.useEffect(() => {
-    if (portalContainer.closest('[data-overlay-container]')) {
+    setCurrentPortalContainer(portalContainer ?? document.body);
+  }, [portalContainer]);
+
+  React.useEffect(() => {
+    if (currentPortalContainer?.closest('[data-overlay-container]')) {
       throw new Error('An OverlayContainer must not be inside another container. Please change the portalContainer prop.');
     }
-  }, [portalContainer]);
+  }, [currentPortalContainer]);
+
+  if (!currentPortalContainer) {
```

Comment:
Why do we need state for this? Won't that introduce an extra render? Could we change the above to `portalContainer = typeof document !== 'undefined' ? document.body : null`, and this to `if (!portalContainer)` instead?

Reply:
This approach seems to work well, but SSR test did not pass because the results of the server render and the client's initial render are different.

```
Warning: Expected server HTML to contain a matching <div> in <div>.
```

ref: https://app.circleci.com/pipelines/github/adobe/react-spectrum/9241/workflows/addeabee-d62d-4f5c-b78a-892173ba46ba/jobs/89756

Reply:
Ah ok. Then I think we can use useIsSSR for this in addition to my previously proposed solution. That will cause a re-render only after hydration, so no extra re-render for client only apps.

Reply:
https://github.com/adobe/react-spectrum/pull/3186/commits/e68a26ff8c89012909d3f40d44a40869904ca1f5 ✅